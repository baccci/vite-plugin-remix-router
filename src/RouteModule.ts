import path from 'path'
import { RouteObject } from 'react-router-dom'
import type { RouteNode } from './buildRouteTree'
import { isDirectory, normalizeFilenameToRoute } from './utils'
import { getOptions } from './options'

function createRouteElement(filePath: string) {
  return `::React.createElement(React.lazy(() => import("/${filePath}")))::`
}

export class RouteModule {
  private routes: RouteObject | undefined

  private imports: Array<string> = []

  private absolutePath(filePath: string) {
    return path.join(getOptions().root, filePath)
  }

  buildRouteObject(rootNode: RouteNode) {
    this.imports = []
    this.routes = this.createRouteObject(rootNode)

    return this.routes
  }

  private createRouteObject(node: RouteNode) {
    if (isDirectory(this.absolutePath(node.path))) {
      return this.createLayoutRoute(node)
    }

    return this.createPageRoute(node)
  }

  private createLayoutRoute(node: RouteNode): RouteObject {
    const layout = node.children.find((child) => child.name === '_layout')

    const element = layout ? { element: createRouteElement(layout.path) } : {}

    return {
      ...element,
      path: normalizeFilenameToRoute(node.name),
      children: node.children
        .filter((child) => child.name !== '_layout')
        .map((child) => this.createRouteObject(child)),
    }
  }

  private createPageRoute(node: RouteNode): RouteObject {
    const path =
      node.name === 'index'
        ? { index: true }
        : { path: normalizeFilenameToRoute(node.name) }

    return {
      ...path,
      element: createRouteElement(node.path),
    }
  }

  generate() {
    const code: Array<string> = []
    code.push("import React from 'react';")
    code.push(...this.imports)
    code.push('')

    const routesString = JSON.stringify(this.routes, null, 2)
      .replace(/\\"/g, '"') // check this
      .replace(/("::|::")/g, '')

    code.push(`export const routes = [${routesString}]\n`)

    return code.join('\n')
  }
}
