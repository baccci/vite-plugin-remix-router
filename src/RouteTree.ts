import fs from 'fs'
import path from 'path'
import { ResolvedOptions } from './types'
import { isDirectory } from './utils'

export class RouteNode {
  name: string

  path: string

  children: Array<RouteNode> = []

  constructor(filePath: string) {
    this.name = path.parse(filePath).name
    this.path = filePath
  }
}

export class RouteTree {
  private options: ResolvedOptions

  private _rootNode: RouteNode | undefined

  constructor(options: ResolvedOptions) {
    this.options = options
  }

  get root(): RouteNode {
    if (!this._rootNode) {
      return this.build()
    }

    return this._rootNode
  }

  build() {
    this._rootNode = this.createNode(this.options.pageDir)
    this._rootNode.name = '/'

    return this._rootNode
  }

  rebuild() {
    this._rootNode = undefined
    return this.build()
  }

  private createNode(filePath: string) {
    const node = new RouteNode(filePath)

    if (isDirectory(this.absolutePath(filePath))) {
      const children = this.resolveChildren(filePath)

      node.children = children.map((child) =>
        this.createNode(`${filePath}/${child}`),
      )
    }

    return node
  }

  private absolutePath(filePath: string) {
    return path.join(this.options.root, filePath)
  }

  private resolveChildren(directoryPath: string) {
    const children = fs.readdirSync(this.absolutePath(directoryPath))

    return children.filter((child) => {
      const childPath = path.join(directoryPath, child)

      if (isDirectory(childPath)) {
        return true
      }

      const extension = path.extname(childPath)
      return this.options.extensions.includes(extension.substring(1))
    })
  }
}
