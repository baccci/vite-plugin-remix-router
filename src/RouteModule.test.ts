import { RouteModule } from './RouteModule'
import { RouteNode, RouteTree } from './RouteTree'

const options = {
  root: process.cwd(),
  extensions: ['tsx', 'jsx'],
  pageDir: 'example/src/pages',
}

const routeModuleGenerator = new RouteModule(options)

it('should convert RouteNode to RouteObject', () => {
  const routeNode = new RouteNode('example/src/pages/about.tsx')
  const route = routeModuleGenerator.buildRouteObject(routeNode)

  expect(route).toEqual({
    path: 'about',
    element: `::React.createElement(React.lazy(() => import("/${routeNode.path}")))::`,
    getRouteConfig: '::example_src_pages_about_ROUTE_CONFIG::',
  })
})

it('should convert directory RouteNode to RouteObject', () => {
  const routeNode = new RouteNode('example/src/pages/posts')
  routeNode.children = [
    new RouteNode('example/src/pages/posts/[slug].tsx'),
    new RouteNode('example/src/pages/posts/index.tsx'),
  ]

  const { children, ...route } =
    routeModuleGenerator.buildRouteObject(routeNode)

  expect(route).toEqual({
    path: 'posts',
  })
  expect(children).toHaveLength(2)
})

it('should convert dynamic paths to route parameters', () => {
  const routeNode = new RouteNode('example/src/pages/posts/[slug].tsx')
  const route = routeModuleGenerator.buildRouteObject(routeNode)

  expect(route.path).toEqual(':slug')
  expect(route.children).toBeUndefined()
  expect(route.element).toEqual(
    `::React.createElement(React.lazy(() => import("/${routeNode.path}")))::`,
  )
})

it('should convert [...all] routes to `no-match` route', () => {
  const routeNode = new RouteNode('example/src/pages/[...all].tsx')
  const route = routeModuleGenerator.buildRouteObject(routeNode)

  expect(route).toEqual({
    path: '*',
    element: `::React.createElement(React.lazy(() => import("/${routeNode.path}")))::`,
  })
})

it('should create index route for RouteNodes with name `index`', () => {
  const routeNode = new RouteNode('example/src/pages/users/index.tsx')
  const route = routeModuleGenerator.buildRouteObject(routeNode)

  expect(route).toEqual({
    index: true,
    element: `::React.createElement(React.lazy(() => import("/${routeNode.path}")))::`,
  })
})

it('should user `_layout` as element for directories', () => {
  const routeNode = new RouteNode('example/src/pages/users')
  routeNode.children = [
    new RouteNode('example/src/pages/users/_layout.tsx'),
    new RouteNode('example/src/pages/users/index.tsx'),
  ]

  const { children, ...route } =
    routeModuleGenerator.buildRouteObject(routeNode)

  expect(children).toHaveLength(1)
  expect(route).toEqual({
    path: 'users',
    element: `::React.createElement(React.lazy(() => import("/${routeNode.path}/_layout.tsx")))::`,
    getRouteConfig: '::example_src_pages_users__layout_ROUTE_CONFIG::',
  })
})

it('should add import for page getRouteConfig if exists', () => {
  const routeNode = new RouteNode('example/src/pages/about.tsx')
  routeModuleGenerator.buildRouteObject(routeNode)
  const routeModule = routeModuleGenerator.generate()

  expect(routeModule).toMatchInlineSnapshot(`
    "import React from 'react';
    import { getRouteConfig as example_src_pages_about_ROUTE_CONFIG } from '/example/src/pages/about.tsx';
    
    export const routes = [{
      \\"path\\": \\"about\\",
      \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/about.tsx\\"))),
      \\"getRouteConfig\\": example_src_pages_about_ROUTE_CONFIG
    }]
    
    export function traverse(routes, callback) {
      return routes.map((route) => {
        const modifiedRouteObject = callback(route);
        if (route.children) {
          modifiedRouteObject.children = traverse(route.children, callback);
        }
        return modifiedRouteObject;
      });
    }"
  `)
})

it('should add import for `_layout` getRouteConfig if exists', () => {
  const routeNode = new RouteNode('example/src/pages/users')
  routeNode.children = [
    new RouteNode('example/src/pages/users/_layout.tsx'),
    new RouteNode('example/src/pages/users/index.tsx'),
  ]

  routeModuleGenerator.buildRouteObject(routeNode)
  const routeModule = routeModuleGenerator.generate()

  expect(routeModule).toMatchInlineSnapshot(`
    "import React from 'react';
    import { getRouteConfig as example_src_pages_users__layout_ROUTE_CONFIG } from '/example/src/pages/users/_layout.tsx';
    
    export const routes = [{
      \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/users/_layout.tsx\\"))),
      \\"path\\": \\"users\\",
      \\"getRouteConfig\\": example_src_pages_users__layout_ROUTE_CONFIG,
      \\"children\\": [
        {
          \\"index\\": true,
          \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/users/index.tsx\\")))
        }
      ]
    }]
    
    export function traverse(routes, callback) {
      return routes.map((route) => {
        const modifiedRouteObject = callback(route);
        if (route.children) {
          modifiedRouteObject.children = traverse(route.children, callback);
        }
        return modifiedRouteObject;
      });
    }"
  `)
})

it('should generate routes module', () => {
  const routeNode = new RouteNode('example/src/pages/about.tsx')
  routeModuleGenerator.buildRouteObject(routeNode)
  const routeModule = routeModuleGenerator.generate()

  expect(routeModule).toMatchInlineSnapshot(`
    "import React from 'react';
    import { getRouteConfig as example_src_pages_about_ROUTE_CONFIG } from '/example/src/pages/about.tsx';
    
    export const routes = [{
      \\"path\\": \\"about\\",
      \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/about.tsx\\"))),
      \\"getRouteConfig\\": example_src_pages_about_ROUTE_CONFIG
    }]
    
    export function traverse(routes, callback) {
      return routes.map((route) => {
        const modifiedRouteObject = callback(route);
        if (route.children) {
          modifiedRouteObject.children = traverse(route.children, callback);
        }
        return modifiedRouteObject;
      });
    }"
  `)
})

it('should match routes snapshot', () => {
  const routeTree = new RouteTree(options).build()
  const routes = routeModuleGenerator.buildRouteObject(routeTree)
  expect(routes).toMatchSnapshot()
})

it('should match route modules snapshot', () => {
  const routeTree = new RouteTree(options).build()
  routeModuleGenerator.buildRouteObject(routeTree)
  const routesModule = routeModuleGenerator.generate()
  expect(routesModule).toMatchSnapshot()
})
