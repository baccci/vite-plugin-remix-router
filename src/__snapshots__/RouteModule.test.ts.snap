// Vitest Snapshot v1

exports[`should match route modules snapshot 1`] = `
"import React from 'react';

export const routes = [{
  \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/_layout.tsx\\"))),
  \\"path\\": \\"/\\",
  \\"children\\": [
    {
      \\"path\\": \\"*\\",
      \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/$.tsx\\")))
    },
    {
      \\"path\\": \\"about\\",
      \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/about.tsx\\")))
    },
    {
      \\"index\\": true,
      \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/index.tsx\\")))
    },
    {
      \\"path\\": \\"posts\\",
      \\"children\\": [
        {
          \\"path\\": \\":slug\\",
          \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/posts/[slug].tsx\\")))
        },
        {
          \\"index\\": true,
          \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/posts/index.tsx\\")))
        }
      ]
    },
    {
      \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/users/_layout.tsx\\"))),
      \\"path\\": \\"users\\",
      \\"children\\": [
        {
          \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/users/[user]/_layout.tsx\\"))),
          \\"path\\": \\":user\\",
          \\"children\\": [
            {
              \\"index\\": true,
              \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/users/[user]/index.tsx\\")))
            },
            {
              \\"path\\": \\"profile\\",
              \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/users/[user]/profile.tsx\\")))
            },
            {
              \\"path\\": \\"settings\\",
              \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/users/[user]/settings.tsx\\")))
            }
          ]
        },
        {
          \\"index\\": true,
          \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/users/index.tsx\\")))
        }
      ]
    }
  ]
}]
"
`;

exports[`should match routes snapshot 1`] = `
{
  "children": [
    {
      "element": "::React.createElement(React.lazy(() => import(\\"/example/src/pages/$.tsx\\")))::",
      "path": "*",
    },
    {
      "element": "::React.createElement(React.lazy(() => import(\\"/example/src/pages/about.tsx\\")))::",
      "path": "about",
    },
    {
      "element": "::React.createElement(React.lazy(() => import(\\"/example/src/pages/index.tsx\\")))::",
      "index": true,
    },
    {
      "children": [
        {
          "element": "::React.createElement(React.lazy(() => import(\\"/example/src/pages/posts/[slug].tsx\\")))::",
          "path": ":slug",
        },
        {
          "element": "::React.createElement(React.lazy(() => import(\\"/example/src/pages/posts/index.tsx\\")))::",
          "index": true,
        },
      ],
      "path": "posts",
    },
    {
      "children": [
        {
          "children": [
            {
              "element": "::React.createElement(React.lazy(() => import(\\"/example/src/pages/users/[user]/index.tsx\\")))::",
              "index": true,
            },
            {
              "element": "::React.createElement(React.lazy(() => import(\\"/example/src/pages/users/[user]/profile.tsx\\")))::",
              "path": "profile",
            },
            {
              "element": "::React.createElement(React.lazy(() => import(\\"/example/src/pages/users/[user]/settings.tsx\\")))::",
              "path": "settings",
            },
          ],
          "element": "::React.createElement(React.lazy(() => import(\\"/example/src/pages/users/[user]/_layout.tsx\\")))::",
          "path": ":user",
        },
        {
          "element": "::React.createElement(React.lazy(() => import(\\"/example/src/pages/users/index.tsx\\")))::",
          "index": true,
        },
      ],
      "element": "::React.createElement(React.lazy(() => import(\\"/example/src/pages/users/_layout.tsx\\")))::",
      "path": "users",
    },
  ],
  "element": "::React.createElement(React.lazy(() => import(\\"/example/src/pages/_layout.tsx\\")))::",
  "path": "/",
}
`;
